azerothcore/azerothcore-wotlk
# (feature) Add `Rate.Drop.Item.GroupAmount` to increase loot generated from loot groups

## Changes Proposed:
-  adds a `Rate.Drop.Item.GroupAmount` setting to `worldserver.conf`, defaults to `1`.
- `Rate.Drop.Item.GroupAmount` changes the amount of times loot is created from loot groups (Blizzlike is 1).
- Will only affect loot groups in the top-level loot tables, does not cascade down to any reference loot tables, nor does it affect reference links that are in groups.
- Affects all kinds of loot that uses loot groups (creatures, items, etc)
- Does not affect quest items.

Allows you to increase drops from loot tables that use the `GroupId` field when defining items. In Blizzlike, this would allow only one item from the loot group to be selected. Changing `Rate.Drop.Item.GroupAmount` to a number higher than 1 causes multiple items to be selected.

Serves much the same functionality as the existing `Rate.Drop.Item.ReferencedAmount` which only multiplies referenced loot. This new setting does not conflict with `Rate.Drop.Item.ReferencedAmount` nor does it result in absurd amounts of loot if they are used together.

## Issues Addressed:
None I could find.

## SOURCE:
Personal testing.

## Tests Performed:

- clean build
- tested with enemies with various loot list configurations (only explicitly-defined, explicitly-defined with groups, with and without references)

Good test enemies:
- 6487 (Arcanist Doan) - one group, no references. The new setting multiplies the loot from the one group.
- 379 (Darcy) - no groups, no references. The new setting has NO effect.
- 462 (Vultros) - three groups, no references. The new setting multiplies the amount of loot received from the groups. He is a rare, so you will need to spawn him. Note that groups are not guaranteed to hit every time, but each time they do hit you'll get `Rate.Drop.Item.GroupAmount` items from each group.
- 12118 (Lucifron) - two guaranteed references, one 10% chance reference, no groups. The new setting has NO effect (drops one item from each of the two guaranteed references). Notably the groups inside those references are NOT affected.

Good test items:
- 27481 (Heavy Supply Crate) - one group, no references. New setting multiplies the loot from the one group.
- 17965 (Abandoned Adventurer's Satchel) - no groups, no references. New setting has NO effect.
- 46007 (Bag of Fishing Treasures) - no groups, four references. New setting has NO effect.

## How to Test the Changes:

1. Fresh server, fresh character, default settings.
2. `.gm on`
3. `.list creature 6487` to find the GUID for Arcanist Doan
4. `.go creature <GUID_from_above>`
5. `/target arcanist doan`, `.die`
6. Check loot and observe that only one blue/epic item is dropped. This item was selected from loot `GroupId` 1.
7. Update `Rate.Drop.Item.GroupAmount` in `worldserver.conf` to `3`
8. `.reload config`
9. `/target arcanist doan`, `.respawn`
10. `/target arcanist doan`, `.die`
11. Check loot and observe that three blue/epic items are dropped.

## Known Issues and TODO List:

- Perhaps update the code to only exclude quest items if they're non-stackable.
- Alter the logic for non-guaranteed groups to have more chances to get one of the group's items. Current functionality will emit `Rate.Drop.Item.GroupAmount` items from each group that is rolled for only once.
- Create configuration option for the user to select if they want to allow duplicate items or not.


## How to Test AzerothCore PRs
 
When a PR is ready to be tested, it will be marked as **[WAITING TO BE TESTED]**.

You can help by testing PRs and writing your feedback here on the PR's page on GitHub. Follow the instructions here:

http://www.azerothcore.org/wiki/How-to-test-a-PR

**REMEMBER**: when testing a PR that changes something **generic** (i.e. a part of code that handles more than one specific thing), the tester should not only check that the PR does its job (e.g. fixing spell XXX) but **especially** check that the PR does not cause any regression (i.e. introducing new bugs).

**For example**: if a PR fixes spell X by changing a part of code that handles spells X, Y, and Z, we should not only test X, but **we should test Y and Z as well**.
